File: paimon-common/src/main/java/org/apache/paimon/globalindex/SortedFileGlobalIndexReader.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.globalindex;

import org.apache.paimon.fs.Path;
import org.apache.paimon.predicate.Contains;
import org.apache.paimon.predicate.EndsWith;
import org.apache.paimon.predicate.Equal;
import org.apache.paimon.predicate.FieldRef;
import org.apache.paimon.predicate.LeafBinaryFunction;
import org.apache.paimon.predicate.Like;
import org.apache.paimon.predicate.LikeOptimization;
import org.apache.paimon.predicate.StartsWith;
import org.apache.paimon.types.DataTypeFamily;
import org.apache.paimon.utils.Pair;
import org.apache.paimon.utils.RoaringNavigableMap64;

import java.io.Closeable;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.function.Function;
import java.util.function.Supplier;

/** Base reader for sorted global index files with manifest-level min/max pruning. */
public abstract class SortedFileGlobalIndexReader<R extends Closeable>
        implements GlobalIndexReader {

    private final List<GlobalIndexIOMeta> files;
    private final SortedFileMetaSelector fileSelector;
    private final long fallbackScanMaxSize;
    private final Map<Path, R> readerCache;
    private final ExecutorService executor;

    protected SortedFileGlobalIndexReader(
            List<GlobalIndexIOMeta> files,
            KeySerializer keySerializer,
            long fallbackScanMaxSize,
            ExecutorService executor) {
        this.files = files;
        this.fileSelector = new SortedFileMetaSelector(files, keySerializer);
        this.fallbackScanMaxSize = fallbackScanMaxSize;
        this.readerCache = new ConcurrentHashMap<>();
        this.executor = executor;
    }

    /** Fans out {@code visitor} over every file of this range and unions the per-file results. */
    protected CompletableFuture<Optional<GlobalIndexResult>> visitAllFiles(
            Function<R, Optional<GlobalIndexResult>> visitor) {
        return visitSelectedFiles(Optional.of(files), visitor);
    }

    /**
     * Conservative "all non-null rows" answer shared by the string negations that cannot prune
     * (ENDS WITH / CONTAINS / unoptimizable LIKE). The default is a budget-gated fallback scan;
     * subclasses that can derive it cheaply may override.
     */
    protected CompletableFuture<Optional<GlobalIndexResult>> visitConservativeString(
            FieldRef fieldRef,
            Object literal,
            Supplier<Optional<List<GlobalIndexIOMeta>>> selector,
            Function<R, Optional<GlobalIndexResult>> visitor) {
        if (!canFallbackStringScan(fieldRef, literal)) {
            return unsupported();
        }
        return visitFallbackParallel(selector, visitor);
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitIsNotNull(FieldRef fieldRef) {
        return visitParallel(() -> fileSelector.visitIsNotNull(fieldRef), this::visitIsNotNull);
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitIsNull(FieldRef fieldRef) {
        return visitParallel(() -> fileSelector.visitIsNull(fieldRef), this::visitIsNull);
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitStartsWith(
            FieldRef fieldRef, Object literal) {
        if (!fieldRef.type().is(DataTypeFamily.CHARACTER_STRING) || literal == null) {
            return unsupported();
        }
        return visitParallel(
                () -> fileSelector.visitStartsWith(fieldRef, literal),
                reader -> visitStartsWith(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitEndsWith(
            FieldRef fieldRef, Object literal) {
        return visitConservativeString(
                fieldRef,
                literal,
                () -> fileSelector.visitEndsWith(fieldRef, literal),
                reader -> visitEndsWith(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitContains(
            FieldRef fieldRef, Object literal) {
        return visitConservativeString(
                fieldRef,
                literal,
                () -> fileSelector.visitContains(fieldRef, literal),
                reader -> visitContains(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitLike(
            FieldRef fieldRef, Object literal) {
        if (!fieldRef.type().is(DataTypeFamily.CHARACTER_STRING) || literal == null) {
            return unsupported();
        }

        Optional<Pair<LeafBinaryFunction, Object>> optimized =
                LikeOptimization.tryOptimize(literal);
        if (!optimized.isPresent()) {
            return visitConservativeString(
                    fieldRef,
                    literal,
                    () -> fileSelector.visitLike(fieldRef, literal),
                    reader -> visitLike(reader, fieldRef, literal));
        }

        LeafBinaryFunction function = optimized.get().getKey();
        Object optimizedLiteral = optimized.get().getValue();
        if (function == Equal.INSTANCE) {
            return visitEqual(fieldRef, optimizedLiteral);
        }
        if (function == StartsWith.INSTANCE) {
            return visitStartsWith(fieldRef, optimizedLiteral);
        }
        if (function == EndsWith.INSTANCE) {
            return visitEndsWith(fieldRef, optimizedLiteral);
        }
        if (function == Contains.INSTANCE) {
            return visitContains(fieldRef, optimizedLiteral);
        }
        return unsupported();
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitLessThan(
            FieldRef fieldRef, Object literal) {
        if (!canFallbackScan(literal)) {
            return unsupported();
        }
        return visitFallbackParallel(
                () -> fileSelector.visitLessThan(fieldRef, literal),
                reader -> visitLessThan(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitGreaterOrEqual(
            FieldRef fieldRef, Object literal) {
        if (!canFallbackScan(literal)) {
            return unsupported();
        }
        return visitFallbackParallel(
                () -> fileSelector.visitGreaterOrEqual(fieldRef, literal),
                reader -> visitGreaterOrEqual(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitNotEqual(
            FieldRef fieldRef, Object literal) {
        return visitParallel(
                () -> fileSelector.visitNotEqual(fieldRef, literal),
                reader -> visitNotEqual(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitLessOrEqual(
            FieldRef fieldRef, Object literal) {
        if (!canFallbackScan(literal)) {
            return unsupported();
        }
        return visitFallbackParallel(
                () -> fileSelector.visitLessOrEqual(fieldRef, literal),
                reader -> visitLessOrEqual(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitEqual(
            FieldRef fieldRef, Object literal) {
        return visitParallel(
                () -> fileSelector.visitEqual(fieldRef, literal),
                reader -> visitEqual(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitGreaterThan(
            FieldRef fieldRef, Object literal) {
        if (!canFallbackScan(literal)) {
            return unsupported();
        }
        return visitFallbackParallel(
                () -> fileSelector.visitGreaterThan(fieldRef, literal),
                reader -> visitGreaterThan(reader, literal));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitIn(
            FieldRef fieldRef, List<Object> literals) {
        return visitParallel(
                () -> fileSelector.visitIn(fieldRef, literals),
                reader -> visitIn(reader, literals));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitNotIn(
            FieldRef fieldRef, List<Object> literals) {
        return visitParallel(
                () -> fileSelector.visitNotIn(fieldRef, literals),
                reader -> visitNotIn(reader, literals));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitBetween(
            FieldRef fieldRef, Object from, Object to) {
        if (!canFallbackScan(from) || to == null) {
            return unsupported();
        }
        return visitFallbackParallel(
                () -> fileSelector.visitBetween(fieldRef, from, to),
                reader -> visitBetween(reader, from, to));
    }

    @Override
    public CompletableFuture<Optional<GlobalIndexResult>> visitNotBetween(
            FieldRef fieldRef, Object from, Object to) {
        if (!canFallbackScan(from) || to == null) {
            return unsupported();
        }
        return visitFallbackParallel(
                () ->
                        fileSelector.visitOr(
                                Arrays.asList(
                                        fileSelector.visitLessThan(fieldRef, from),
                                        fileSelector.visitGreaterThan(fieldRef, to))),
                reader -> visitNotBetween(reader, from, to));
    }

    protected abstract Optional<GlobalIndexResult> visitIsNotNull(R reader);

    protected abstract Optional<GlobalIndexResult> visitIsNull(R reader);

    protected abstract Optional<GlobalIndexResult> visitStartsWith(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitEndsWith(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitContains(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitLessThan(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitGreaterOrEqual(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitNotEqual(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitLessOrEqual(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitEqual(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitGreaterThan(R reader, Object literal);

    protected abstract Optional<GlobalIndexResult> visitIn(R reader, List<Object> literals);

    protected abstract Optional<GlobalIndexResult> visitNotIn(R reader, List<Object> literals);

    protected abstract Optional<GlobalIndexResult> visitBetween(R reader, Object from, Object to);

    protected Optional<GlobalIndexResult> visitLike(R reader, FieldRef fieldRef, Object literal) {
        return createResult(like(reader, key -> Like.INSTANCE.test(fieldRef.type(), key, literal)));
    }

    protected Optional<GlobalIndexResult> visitNotBetween(R reader, Object from, Object to) {
        RoaringNavigableMap64 result = lessThan(reader, from);
        result.or(greaterThan(reader, to));
        return Optional.of(GlobalIndexResult.create(result));
    }

    protected RoaringNavigableMap64 like(R reader, Function<Object, Boolean> keyPredicate) {
        throw new UnsupportedOperationException();
    }

    protected abstract RoaringNavigableMap64 lessThan(R reader, Object literal);

    protected abstract RoaringNavigableMap64 greaterThan(R reader, Object literal);

    protected Optional<GlobalIndexResult> createResult(RoaringNavigableMap64 bitmap) {
        return Optional.of(GlobalIndexResult.create(bitmap));
    }

    protected abstract R openReader(GlobalIndexIOMeta meta);

    @Override
    public void close() throws IOException {
        IOException exception = null;
        for (R reader : readerCache.values()) {
            try {
                reader.close();
            } catch (IOException e) {
                if (exception == null) {
                    exception = e;
                } else {
                    exception.addSuppressed(e);
                }
            }
        }
        if (exception != null) {
            throw exception;
        }
    }

    private boolean canFallbackStringScan(FieldRef fieldRef, Object literal) {
        return fallbackScanMaxSize > 0
                && fieldRef.type().is(DataTypeFamily.CHARACTER_STRING)
                && literal != null;
    }

    private boolean canFallbackScan(Object literal) {
        return fallbackScanMaxSize > 0 && literal != null;
    }

    private static boolean fallbackScanEnabled(List<GlobalIndexIOMeta> files, long maxSize) {
        if (maxSize <= 0) {
            return false;
        }
        long totalSize = 0;
        for (GlobalIndexIOMeta file : files) {
            if (Long.MAX_VALUE - totalSize < file.fileSize()) {
                return false;
            }
            totalSize += file.fileSize();
            if (totalSize > maxSize) {
                return false;
            }
        }
        return true;
    }

    private CompletableFuture<Optional<GlobalIndexResult>> unsupported() {
        return CompletableFuture.completedFuture(Optional.empty());
    }

    private CompletableFuture<Optional<GlobalIndexResult>> visitParallel(
            Supplier<Optional<List<GlobalIndexIOMeta>>> selector,
            Function<R, Optional<GlobalIndexResult>> visitor) {
        return visitSelectedFiles(selector.get(), visitor);
    }

    private CompletableFuture<Optional<GlobalIndexResult>> visitFallbackParallel(
            Supplier<Optional<List<GlobalIndexIOMeta>>> selector,
            Function<R, Optional<GlobalIndexResult>> visitor) {
        Optional<List<GlobalIndexIOMeta>> selectedOpt = selector.get();
        if (!selectedOpt.isPresent()) {
            return CompletableFuture.completedFuture(Optional.empty());
        }
        List<GlobalIndexIOMeta> selected = selectedOpt.get();
        if (selected.isEmpty()) {
            return CompletableFuture.completedFuture(Optional.of(GlobalIndexResult.createEmpty()));
        }
        if (!fallbackScanEnabled(selected, fallbackScanMaxSize)) {
            return unsupported();
        }
        return visitSelectedFiles(selectedOpt, visitor);
    }

    private CompletableFuture<Optional<GlobalIndexResult>> visitSelectedFiles(
            Optional<List<GlobalIndexIOMeta>> selectedOpt,
            Function<R, Optional<GlobalIndexResult>> visitor) {
        if (!selectedOpt.isPresent()) {
            return unsupported();
        }
        List<GlobalIndexIOMeta> selected = selectedOpt.get();
        if (selected.isEmpty()) {
            return CompletableFuture.completedFuture(Optional.of(GlobalIndexResult.createEmpty()));
        }

        List<CompletableFuture<Optional<GlobalIndexResult>>> futures =
                new ArrayList<>(selected.size());
        for (GlobalIndexIOMeta meta : selected) {
            futures.add(
                    CompletableFuture.supplyAsync(
                            () -> visitor.apply(getOrCreateReader(meta)), executor));
        }
        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
                .thenApply(v -> unionResults(futures));
    }

    private R getOrCreateReader(GlobalIndexIOMeta meta) {
        return readerCache.computeIfAbsent(meta.filePath(), ignored -> openReader(meta));
    }

    private Optional<GlobalIndexResult> unionResults(
            List<CompletableFuture<Optional<GlobalIndexResult>>> futures) {
        Optional<GlobalIndexResult> result = Optional.empty();
        for (CompletableFuture<Optional<GlobalIndexResult>> future : futures) {
            Optional<GlobalIndexResult> current = future.join();
            if (!current.isPresent()) {
                continue;
            }
            result = result.isPresent() ? Optional.of(result.get().or(current.get())) : current;
        }
        return result;
    }
}

File: paimon-python/pypaimon/globalindex/sorted_file_global_index_reader.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""Base reader for sorted global index files with manifest pruning."""

import threading
from concurrent.futures import Executor, Future
from typing import Callable, Dict, List, Optional

from pypaimon.globalindex.global_index_meta import GlobalIndexIOMeta
from pypaimon.globalindex.global_index_reader import FieldRef, GlobalIndexReader, _completed_future
from pypaimon.globalindex.global_index_result import GlobalIndexResult
from pypaimon.globalindex.key_serializer import KeySerializer
from pypaimon.globalindex.sorted_file_meta_selector import SortedFileMetaSelector
from pypaimon.globalindex.sorted_index_file_meta import SortedIndexFileMeta
from pypaimon.utils.roaring_bitmap import RoaringBitmap64


_LONG_MAX_VALUE = (1 << 63) - 1


class SortedFileGlobalIndexReader(GlobalIndexReader):
    """Shared lazy reader/candidate pruning logic for sorted global indexes."""

    def __init__(
        self,
        key_serializer: KeySerializer,
        file_io,
        index_path: str,
        io_metas: List[GlobalIndexIOMeta],
        executor: Executor,
        fallback_scan_max_size: int,
    ):
        self._key_serializer = key_serializer
        self._file_io = file_io
        self._index_path = index_path
        self._executor = executor
        self._fallback_scan_max_size = fallback_scan_max_size

        files = []
        for io_meta in io_metas:
            idx_meta = SortedIndexFileMeta.deserialize(io_meta.metadata)
            files.append((io_meta, idx_meta))

        self._selector = SortedFileMetaSelector(files, key_serializer)
        self._reader_cache: Dict[str, object] = {}
        self._cache_lock = threading.Lock()

    def open_reader(self, meta: GlobalIndexIOMeta):
        raise NotImplementedError

    def _get_or_create_reader(self, meta: GlobalIndexIOMeta):
        key = meta.external_path or meta.file_name
        with self._cache_lock:
            reader = self._reader_cache.get(key)
            if reader is not None:
                return reader
            reader = self.open_reader(meta)
            self._reader_cache[key] = reader
            return reader

    def _visit_parallel(
        self,
        selector_fn: Callable[[], Optional[List[GlobalIndexIOMeta]]],
        visitor_fn: Callable[[object], Optional[GlobalIndexResult]],
    ) -> 'Future[Optional[GlobalIndexResult]]':
        selected = selector_fn()
        if selected is None:
            return _completed_future(None)
        if not selected:
            return _completed_future(GlobalIndexResult.create_empty())

        task_futures: List[Future] = []
        for meta in selected:
            task_futures.append(self._executor.submit(
                lambda m=meta: visitor_fn(self._get_or_create_reader(m))))

        all_done: Future = Future()
        remaining = [len(task_futures)]
        lock = threading.Lock()

        def on_done(_):
            with lock:
                remaining[0] -= 1
                if remaining[0] == 0:
                    try:
                        result: Optional[GlobalIndexResult] = None
                        for future in task_futures:
                            current = future.result()
                            if current is None:
                                continue
                            if result is None:
                                result = current
                            else:
                                result = result.or_(current)
                        all_done.set_result(result)
                    except Exception as e:
                        all_done.set_exception(e)

        for future in task_futures:
            future.add_done_callback(on_done)

        return all_done

    def _visit_fallback_parallel(
        self,
        selector_fn: Callable[[], Optional[List[GlobalIndexIOMeta]]],
        visitor_fn: Callable[[object], Optional[GlobalIndexResult]],
    ) -> 'Future[Optional[GlobalIndexResult]]':
        selected = selector_fn()
        if selected is None:
            return _completed_future(None)
        if not selected:
            return _completed_future(GlobalIndexResult.create_empty())
        if not self._fallback_scan_enabled(selected):
            return _completed_future(None)
        return self._visit_parallel(lambda: selected, visitor_fn)

    def _fallback_scan_enabled(self, files: List[GlobalIndexIOMeta]) -> bool:
        if self._fallback_scan_max_size <= 0:
            return False
        total_size = 0
        for file in files:
            if _LONG_MAX_VALUE - total_size < file.file_size:
                return False
            total_size += file.file_size
            if total_size > self._fallback_scan_max_size:
                return False
        return True

    def visit_equal(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        return self._visit_parallel(
            lambda: self._selector.select_equal(literal),
            lambda reader: reader.visit_equal(literal))

    def visit_not_equal(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        return self._visit_parallel(
            lambda: self._selector.select_not_equal(literal),
            lambda reader: reader.visit_not_equal(literal))

    def visit_less_than(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._can_fallback_scan(literal):
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_less_than(literal),
            lambda reader: reader.visit_less_than(literal))

    def visit_less_or_equal(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._can_fallback_scan(literal):
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_less_or_equal(literal),
            lambda reader: reader.visit_less_or_equal(literal))

    def visit_greater_than(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._can_fallback_scan(literal):
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_greater_than(literal),
            lambda reader: reader.visit_greater_than(literal))

    def visit_greater_or_equal(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._can_fallback_scan(literal):
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_greater_or_equal(literal),
            lambda reader: reader.visit_greater_or_equal(literal))

    def visit_is_null(self, field_ref: FieldRef) -> 'Future[Optional[GlobalIndexResult]]':
        return self._visit_parallel(
            lambda: self._selector.select_is_null(),
            lambda reader: reader.visit_is_null())

    def visit_is_not_null(self, field_ref: FieldRef) -> 'Future[Optional[GlobalIndexResult]]':
        return self._visit_parallel(
            lambda: self._selector.select_is_not_null(),
            lambda reader: reader.visit_is_not_null())

    def visit_in(self, field_ref: FieldRef, literals) -> 'Future[Optional[GlobalIndexResult]]':
        return self._visit_parallel(
            lambda: self._selector.select_in(literals),
            lambda reader: reader.visit_in(literals))

    def visit_not_in(self, field_ref: FieldRef, literals) -> 'Future[Optional[GlobalIndexResult]]':
        return self._visit_parallel(
            lambda: self._selector.select_not_in(literals),
            lambda reader: reader.visit_not_in(literals))

    def visit_starts_with(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._is_string_type(field_ref) or literal is None:
            return _completed_future(None)
        return self._visit_parallel(
            lambda: self._selector.select_starts_with(literal),
            lambda reader: reader.visit_starts_with(literal))

    def visit_ends_with(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._can_fallback_string_scan(field_ref, literal):
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_ends_with(literal),
            lambda reader: reader.visit_ends_with(literal))

    def visit_contains(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._can_fallback_string_scan(field_ref, literal):
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_contains(literal),
            lambda reader: reader.visit_contains(literal))

    def visit_like(self, field_ref: FieldRef, literal) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._is_string_type(field_ref) or literal is None:
            return _completed_future(None)

        optimized = self._try_optimize_like(literal)
        if optimized is not None:
            method, optimized_literal = optimized
            if method == "equal":
                return self.visit_equal(field_ref, optimized_literal)
            if method == "starts_with":
                return self.visit_starts_with(field_ref, optimized_literal)
            if method == "ends_with":
                return self.visit_ends_with(field_ref, optimized_literal)
            if method == "contains":
                return self.visit_contains(field_ref, optimized_literal)

        if not self._can_fallback_string_scan(field_ref, literal):
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_like(literal),
            lambda reader: reader.visit_like(literal))

    def visit_between(self, field_ref: FieldRef, from_v, to_v) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._can_fallback_scan(from_v) or to_v is None:
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_between(from_v, to_v),
            lambda reader: reader.visit_between(from_v, to_v))

    def visit_not_between(self, field_ref: FieldRef, from_v, to_v) -> 'Future[Optional[GlobalIndexResult]]':
        if not self._can_fallback_scan(from_v) or to_v is None:
            return _completed_future(None)
        return self._visit_fallback_parallel(
            lambda: self._selector.select_or([
                self._selector.select_less_than(from_v),
                self._selector.select_greater_than(to_v),
            ]),
            lambda reader: GlobalIndexResult.create(
                RoaringBitmap64.or_(
                    reader.less_than(from_v),
                    reader.greater_than(to_v))))

    def _can_fallback_string_scan(self, field_ref: FieldRef, literal) -> bool:
        return (self._fallback_scan_max_size > 0
                and literal is not None
                and self._is_string_type(field_ref))

    def _can_fallback_scan(self, literal) -> bool:
        return self._fallback_scan_max_size > 0 and literal is not None

    @staticmethod
    def _is_string_type(field_ref: FieldRef) -> bool:
        data_type = str(getattr(field_ref, "data_type", "")).upper()
        if data_type.endswith(" NOT NULL"):
            data_type = data_type[:-9].rstrip()
        return (data_type == "STRING"
                or data_type == "CHAR"
                or data_type.startswith("CHAR(")
                or data_type == "VARCHAR"
                or data_type.startswith("VARCHAR("))

    @staticmethod
    def _try_optimize_like(literal):
        pattern = str(literal)
        if "_" in pattern:
            return None
        if "%" not in pattern and pattern:
            return "equal", pattern
        if (pattern.startswith("%")
                and pattern.endswith("%")
                and pattern.count("%") == 2
                and pattern[1:-1]):
            return "contains", pattern[1:-1]
        if pattern.startswith("%") and pattern.count("%") == 1 and pattern[1:]:
            return "ends_with", pattern[1:]
        if pattern.endswith("%") and pattern.count("%") == 1 and pattern[:-1]:
            return "starts_with", pattern[:-1]
        return None

    def close(self) -> None:
        with self._cache_lock:
            for reader in self._reader_cache.values():
                try:
                    reader.close()
                except Exception:
                    pass
            self._reader_cache.clear()
